### PR TITLE
4657 method name is inconsistent with the method behavior

### PR DESCRIPTION
**Pull request by epeartree**
the problem:
#4657 

what was done:
-  added base method `getMessageKeyOrDefault` to base class `BaseResourceMessage` with default behavior;
-  added overwrite of the base method in relevant sub classes to return  default value when a key was not specifically set;
-  adjusting json wrapper classes to take advantage of the new method;
-  providing relevant tests;
-  adding changelog.

### CODE REVIEW

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/BaseJsonMessage.java [hunk 1]
```diff
@@ -70,4 +70,9 @@ public void setHeaders(HapiMessageHeaders theHeaders) {
     public String getMessageKeyOrNull() {
 		return null;
     }
+
+	@Nullable
```
Comment (lukedegruchy) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/BaseJsonMessage.java:74]: I don't get this part.  Why are we forwarding to getMessageKeyOrNull()?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceMessage.java [hunk 2]
```diff
@@ -164,10 +164,15 @@ public void setMediaType(String theMediaType) {
 	}
 
 	@Nullable
-	public String getMessageKeyOrNull() {
+	final public String getMessageKeyOrNull() {
 		return myMessageKey;
 	}
 
+	@Nullable
+	public String getMessageKeyOrDefault() {
```
Comment (lukedegruchy) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceMessage.java:172]: Similar comment here about forwarding to getMessageKeyOrNull()

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/ResourceDeliveryJsonMessage.java [hunk 3]
```diff
@@ -66,6 +66,15 @@ public String getMessageKeyOrNull() {
 		return myPayload.getMessageKeyOrNull();
 	}
 
+	@Override
+	@Nullable
+	public String getMessageKeyOrDefault() {
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/ResourceDeliveryJsonMessage.java:71]: So is null in this case the "default"?

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/ResourceModifiedJsonMessage.java [hunk 4]
```diff
@@ -63,6 +63,15 @@ public String getMessageKeyOrNull() {
 		return myPayload.getMessageKeyOrNull();
 	}
 
+	@Override
+	@Nullable
+	public String getMessageKeyOrDefault() {
+		if (myPayload == null) {
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/subscription/model/ResourceModifiedJsonMessage.java:69]: Same question here

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/BaseJsonMessage.java (hunk 1 above)
Comment (epeartree) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/json/BaseJsonMessage.java:74]: `BaseJsonMessage` is a wrapper class and this is the default implementation.  extending classes like `ResourceModifiedMessage` & `ResourceOperationMessage` will provide a payload property and override those 2 method to provide a 'valid' behaviour.  See below.

**Pull request merged**